packages/expo-device/android/src/main/java/expo/modules/device/DeviceModule.java

```diff
@@ -0,0 +1,363 @@
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getIpAddressAsync(Promise promise) {
+    try {
+      Integer ipAddress = getWifiInfo().getIpAddress();
+      // Convert little-endian to big-endianif needed
+      if (ByteOrder.nativeOrder().equals(ByteOrder.LITTLE_ENDIAN)) {
+        ipAddress = Integer.reverseBytes(ipAddress);
+      }
+      byte[] ipByteArray = BigInteger.valueOf(ipAddress).toByteArray();
+      String ipAddressString = InetAddress.getByAddress(ipByteArray).getHostAddress();
+      promise.resolve(ipAddressString);
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void getMACAddressAsync(String interfaceName, Promise promise) {
+    String permission = "android.permission.INTERNET";
+    int res = mContext.checkCallingOrSelfPermission(permission);
+
+    String macAddress = "";
+    if (res == PackageManager.PERMISSION_GRANTED) {
+      try {
+        List<NetworkInterface> interfaces = Collections.list(NetworkInterface.getNetworkInterfaces());
+        for (NetworkInterface intf : interfaces) {
+          if (interfaceName != null) {
+            if (!intf.getName().equalsIgnoreCase(interfaceName)) continue;
+          }
+          byte[] mac = intf.getHardwareAddress();
+          if (mac == null) {
+            macAddress = "";
+          }
+          StringBuilder buf = new StringBuilder();
+          for (byte aMac : mac) {
+            buf.append(String.format("%02X:", aMac));
+          }
+          if (buf.length() > 0) {
+            buf.deleteCharAt(buf.length() - 1);
+          }
+          macAddress = buf.toString();
+          break;
+        }
+      } catch (Exception e) {
+        Log.e(TAG, e.getMessage());
+        promise.reject(e);
+      }
+    }
+    promise.resolve(macAddress);
+  }
+
+  @ExpoMethod
+  public void isAirplaneModeEnabledAsync(Promise promise) {
+    boolean isAirPlaneMode = Settings.Global.getInt(mContext.getContentResolver(), Settings.Global.AIRPLANE_MODE_ON, 0) != 0;
+    promise.resolve(isAirPlaneMode);
+  }
+
+  @ExpoMethod
+  public void hasSystemFeatureAsync(String feature, Promise promise) {
+    try {
+      promise.resolve(mContext.getApplicationContext().getPackageManager().hasSystemFeature(feature));
+    } catch (Exception e) {
+      Log.e(TAG, e.getMessage());
+      promise.reject(e);
+    }
+  }
+
+  @ExpoMethod
+  public void isPinOrFingerprintSetAsync(Promise promise) {
+    KeyguardManager keyguardManager = (KeyguardManager) mContext.getApplicationContext().getSystemService(Context.KEYGUARD_SERVICE); //api 16+
+    promise.resolve(keyguardManager.isKeyguardSecure());
+  }
+
+  @ExpoMethod
+  public void getUserAgentAsync(Promise promise) {
+    String userAgent = System.getProperty("http.agent");
+    promise.resolve(userAgent);
+  }
+
+  @ExpoMethod
+  public void getCarrierAsync(Promise promise) {
+    try {
+      promise.resolve(this.getCarrier());
```

Comment:
I would move `getCarrier` implementation here (in this case I would say that extracting to a separate method is unnecessary), but that's just a suggestion

Reply:
btw `getCarrier` at the moment already handles any exceptions, so the `catch` below is unnecessary